swarm/api/api.go

```diff
@@ -330,13 +332,74 @@ func (self *Api) Get(key storage.Key, path string) (reader storage.LazySectionRe
 		// is set as the hash of the manifest (see swarm/api/manifest.go:NewResourceManifest)
 		//
 		// to avoid taking a performance hit hacking a storage.LazySectionReader to wrap the resource key,
+		// if the resource update is of raw type:
 		// we return a typed error instead. Since for all other purposes this is an invalid manifest,
 		// any normal interfacing code will just see an error fail accordingly.
+		//
+		// if the resource update is of multihash type:
+		// we validate the multihash and retrieve the manifest behind it, and resume normal operations from there
 		if entry.ContentType == ResourceContentType {
 			log.Warn("resource type", "key", key, "hash", entry.Hash)
-			return nil, entry.ContentType, http.StatusOK, &ErrResourceReturn{entry.Hash}
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			rsrc, err := self.resource.LookupLatestByName(ctx, entry.Hash, true, &storage.ResourceLookupParams{})
+			_, rsrcData, err := self.resource.GetContent(entry.Hash)
+			if err != nil {
+				apiGetNotFound.Inc(1)
+				status = http.StatusNotFound
+				log.Warn(fmt.Sprintf("get resource content error: %v", err))
+				return reader, mimeType, status, err
+			}
+			if rsrc.Multihash {
+				var fail bool
+
+				// validate data as multihash
+				decodedMultihash, err := multihash.Decode(rsrcData)
+				if err != nil {
+					fail = true
+				} else if decodedMultihash.Code != multihash.SHA3_256 {
+					err = fmt.Errorf("wrong hash type '%s'", decodedMultihash.Name)
+				}
+				if fail {
+					apiGetInvalid.Inc(1)
+					status = http.StatusUnprocessableEntity
+					log.Warn(fmt.Sprintf("invalid resource multihash: %v", err))
+
+					return reader, mimeType, status, err
+				}
+				key = storage.Key(decodedMultihash.Digest)
+				log.Debug("resouce is multihash", "key", key)
+				// is it correct to increment again here?
+				apiGetCount.Inc(1)
+
+				// get the manifest the multihash digest points to
+				trie, err := loadManifest(self.dpa, key, nil)
+				if err != nil {
+					apiGetNotFound.Inc(1)
```

Comment:
is this necessarily a `not found`?

Reply:
Maybe not, but it mirrors the same error in the beginning of the method retrieving the initial manifest.